src/dlr_relayvm.cc

```diff
@@ -0,0 +1,290 @@
+      tvm::runtime::Module::LoadFromFile(path_->model_lib);
+  std::ifstream relay_ob(path_->relay_executable, std::ios::binary);
+  std::string code_data((std::istreambuf_iterator<char>(relay_ob)),
+                        std::istreambuf_iterator<char>());
+
+  vm_executable_ = std::make_shared<tvm::runtime::Module>(
+      tvm::runtime::vm::Executable::Load(code_data, lib));
+  auto vm = tvm::runtime::make_object<tvm::runtime::vm::VirtualMachine>();
+  vm->LoadExecutable(static_cast<tvm::runtime::vm::Executable *>(
+      const_cast<tvm::runtime::Object *>(vm_executable_->get())));
+  vm_module_ = std::make_shared<tvm::runtime::Module>(tvm::runtime::Module(vm));
+
+  tvm::runtime::PackedFunc init = vm_module_->GetFunction("init");
+  init(static_cast<int>(ctx_.device_type), ctx_.device_id);
+}
+
+void RelayVMModel::LoadMetadata() {
+  LoadJsonFromFile(path_->metadata, metadata_);
+}
+
+void RelayVMModel::FetchInputNodesData() {
+  tvm::runtime::vm::Executable *exec =
+      static_cast<tvm::runtime::vm::Executable *>(
+          const_cast<tvm::runtime::Object *>(vm_executable_->get()));
+  num_inputs_ = exec->GetFunctionArity(ENTRY_FUNCTION);
+  input_names_.resize(num_inputs_);
+  input_types_.resize(num_inputs_);
+  inputs_.resize(num_inputs_);
+  for (int i = 0; i < num_inputs_; i++) {
+    input_names_[i] = exec->GetFunctionParameterName(ENTRY_FUNCTION, i);
+  }
+  try {
+    for (int i = 0; i < num_inputs_; i++) {
+      input_types_[i] = metadata_.at("Model").at("Inputs").at(i).at("dtype");
+    }
+  } catch (nlohmann::json::out_of_range &e) {
+    LOG(ERROR) << e.what();
+    throw dmlc::Error("No Input types metadata found!");
+  }
+}
+
+void RelayVMModel::FetchOutputNodesData() {
+  try {
+    num_outputs_ = metadata_.at("Model").at("Outputs").size();
+  } catch (nlohmann::json::out_of_range &e) {
+    LOG(ERROR) << e.what();
+    throw dmlc::Error("No Output metadata found!");
+  }
+  output_names_.resize(num_outputs_);
+  output_types_.resize(num_outputs_);
+  try {
+    for (int i = 0; i < num_outputs_; i++) {
+      output_names_[i] = metadata_.at("Model").at("Outputs").at(i).at("name");
+      output_types_[i] = metadata_.at("Model").at("Outputs").at(i).at("dtype");
+    }
+  } catch (nlohmann::json::out_of_range &e) {
+    LOG(ERROR) << e.what();
+    throw dmlc::Error("No Output metadata found!");
+  }
+}
+
+const char *RelayVMModel::GetInputName(int index) const {
+  CHECK_LT(index, num_inputs_) << "Input index is out of range.";
+  return input_names_[index].c_str();
+}
+
+const char *RelayVMModel::GetInputType(int index) const {
+  CHECK_LT(index, num_inputs_) << "Input index is out of range.";
+  return input_types_[index].c_str();
+}
+
+const char *RelayVMModel::GetWeightName(int index) const {
+  throw dmlc::Error("Not Implemented!");
+}
+
+std::vector<std::string> RelayVMModel::GetWeightNames() const {
+  throw dmlc::Error("Not Implemented!");
+}
+
+void RelayVMModel::GetInput(const char *name, void *input) {
```

Comment:
const char*, void*

Reply:
this is clang formatted --style=google

Reply:
something is probably wrong there because google style associates *and& to the left

Reply:
https://google.github.io/styleguide/cppguide.html#Pointer_and_Reference_Expressions seems like both are correct. I think it depends on what justification the repo uses. Default seem to be right.

Reply:
I didn't follow all our commits in dlr, but the early ones are left associated, so does TVM. I would suggest we keep consistent.